### ISSUE TITLE
[Proposal] has() as a step modulator.

### ISSUE THREAD

**Issue by okram** [tinkerpop/tinkerpop3#471]
Here is a big internal change that may be useful, may not.

A `Step` implements `HasContainerHolder` if it stores `HasContainers`. E.g. `HasStep`, `TinkerGraphStep`, `Neo4jGraphStep`. Why not just make it such that:

```java
public GraphTraversal<S,E> has(...) {
  if(previousStep instanceof HasContainer) 
    previousStep.addHasContainer(new HasContainer(...));
  else
   this.addStep(new HasStep(new HasContainer(...)));
  return this;
}
```

Why is this cool? 

  . GraphStep can implement `HasContainerHolder`.
  . There is no need for a `TinkerGraphStepStrategy`, `Neo4jGraphStepStrategy`, they simply implement `HasContainerHolder`.
  . `RepeatStep` can implement `HasContainerHolder`. ... see below why this is cool.

```java
g.V().repeat(__.out()).until().has(label,'person')
```

In other words, repeat `out()` until you reach a person.

@mbroecheler @dkuppitz @pietermartin @spmallette

**Comment by mbroecheler** [comment]
Wouldn't the has step be semantically ambiguous in the last example? How do you know if it modulates the repeat or applies after the repeat?

**Comment by okram** [comment]
Yes, Mr. Keen Eye, you are right. Here is my thought on that:

  . if `until()` has no parameters, that means it "wants a `has()`"
  . if `until()` has parameters (e.g. `until(t -> t.blah)`), then it will not accept `has()` and thus, add `HasStep` to the traversal.

However, here is the kicker. Chained `has()`.

```java
g.V().repeat(__.out()).until().has(label,'person').has('age',T.gt,30)
```

"Repeat until you reach a person that is 30+ years old."

If you want `has('age',T.gt,30)` to be a `HasStep`...then here is the ghetto solution:

```java
g.V().repeat(__.out()).until().has(label,'person').identity().has('age',T.gt,30)
```

:)

**Comment by mbroecheler** [comment]
I think this is too confusing. I think a cleaner solution would be allowing an anonymous traversal in until:

```groovy
g.V().repeat(__.out()).until(__.has(label,'person').has('age',T.gt,30))
g.V().repeat(__.out()).until(__.has(label,'person')).has('age',T.gt,30)
```

**Comment by okram** [comment]
Ah yes. That is @dkuppitz recommendation as well. Okay, I will do that.
